gstd/gstd_http.c

```diff
@@ -0,0 +1,406 @@
+static void do_request (SoupServer * server, SoupMessage * msg,
+    GHashTable * query, GstdSession * session, const char *path);
+static void server_callback (SoupServer * server, SoupMessage * msg,
+    const char *path, GHashTable * query, SoupClientContext * context,
+    gpointer data);
+
+static void
+gstd_http_class_init (GstdHttpClass * klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GstdIpcClass *gstdipc_class = GSTD_IPC_CLASS (klass);
+  guint debug_color;
+
+  gstdipc_class->get_option_group =
+      GST_DEBUG_FUNCPTR (gstd_http_init_get_option_group);
+  gstdipc_class->start = GST_DEBUG_FUNCPTR (gstd_http_start);
+  object_class->finalize = gstd_http_finalize;
+  gstdipc_class->stop = GST_DEBUG_FUNCPTR (gstd_http_stop);
+
+  /* Initialize debug category with nice colors */
+  debug_color = GST_DEBUG_FG_BLACK | GST_DEBUG_BOLD | GST_DEBUG_BG_WHITE;
+  GST_DEBUG_CATEGORY_INIT (gstd_http_debug, "gstdhttp", debug_color,
+      "Gstd HTTP category");
+}
+
+static void
+gstd_http_init (GstdHttp * self)
+{
+  GST_INFO_OBJECT (self, "Initializing gstd Http");
+  self->port = GSTD_HTTP_DEFAULT_PORT;
+  self->address = g_strdup (GSTD_HTTP_DEFAULT_ADDRESS);
+  self->server = NULL;
+}
+
+static void
+gstd_http_finalize (GObject * object)
+{
+  GstdHttp *self = GSTD_HTTP (object);
+
+  GST_INFO_OBJECT (object, "Deinitializing gstd HTTP");
+
+  if (self->address) {
+    g_free (self->address);
+  }
+  self->address = NULL;
+
+  G_OBJECT_CLASS (gstd_http_parent_class)->finalize (object);
+}
+
+static SoupStatus
+get_status_code (GstdReturnCode ret)
+{
+  SoupStatus status = SOUP_STATUS_OK;
+
+  if (ret == GSTD_EOK) {
+    status = SOUP_STATUS_OK;
+  } else if (ret == GSTD_BAD_COMMAND || ret == GSTD_NO_RESOURCE) {
+    status = SOUP_STATUS_NOT_FOUND;
+  } else if (ret == GSTD_EXISTING_RESOURCE) {
+    status = SOUP_STATUS_CONFLICT;
+  } else if (ret == GSTD_BAD_VALUE) {
+    status = SOUP_STATUS_NO_CONTENT;
+  } else {
+    status = SOUP_STATUS_BAD_REQUEST;
+  }
+
+  return status;
+}
+
+static GstdReturnCode
+do_get (SoupServer * server, SoupMessage * msg, char **output, const char *path,
+    GstdSession * session)
+{
+  gchar *message = NULL;
+  GstdReturnCode ret = GSTD_EOK;
+
+  g_return_val_if_fail (server, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (msg, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (session, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (output, GSTD_NULL_ARGUMENT);
+
+  message = g_strdup_printf ("read %s", path);
+  ret = gstd_parser_parse_cmd (session, message, output);
+  g_free (message);
+
+  return ret;
+}
+
+static GstdReturnCode
+do_post (SoupServer * server, SoupMessage * msg, char *name,
+    char *description, char **output, const char *path, GstdSession * session)
+{
+  gchar *message = NULL;
+  GstdReturnCode ret = GSTD_EOK;
+
+  g_return_val_if_fail (server, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (msg, GSTD_NULL_ARGUMENT);
+  g_return_val_if_fail (session, GSTD_NULL_ARGUMENT);
```

Comment:
Missing pointer guards. Please extrapolate all comments to the rest of the code.